File: apps/cargo-tracker/src/contexts/billing/infrastructure/services/stub-payment-gateway.ts
import { randomUUID } from 'node:crypto';
import { type Clock, systemClock } from '../../../../shared/infrastructure/clock/clock.js';
import type {
  PaymentConfirmation,
  PaymentGatewayPort,
} from '../../application/outboundservices/acl/payment-gateway-port.js';

/**
 * PaymentGatewayPort のスタブ実装（US23-3）。
 * 決済機関は未接続のため、照会に対して常に入金済み（銀行振込・取引参照を採番）を返す。
 * 実際の決済機関接続（HTTP 契約テストを含む）は運用フェーズで差し替える（ADR-007 スタブ ACL パターン）。
 */
export class StubPaymentGateway implements PaymentGatewayPort {
  constructor(private readonly now: Clock = systemClock) {}

  async confirmPayment(invoiceNumber: string, _amount: number): Promise<PaymentConfirmation | null> {
    return {
      paidAt: this.now(),
      method: 'BANK_TRANSFER',
      transactionReference: `TXN-${invoiceNumber}-${randomUUID().slice(0, 8).toUpperCase()}`,
    };
  }
}

File: docs/article/source/ruby/apps/cargo-tracker/packs/billing/app/infrastructure/billing/infrastructure/stub_payment_gateway.rb
# frozen_string_literal: true

module Billing
  module Infrastructure
    # 決済ゲートウェイの既定スタブ実装（US23・外部決済機関の Faraday/WebMock 導入までの暫定）。
    # MVP は常に入金確認成功として扱う。将来 PaymentGatewayPort の Faraday アダプタへ差し替える。
    class StubPaymentGateway < Domain::PaymentGatewayPort
      def confirm_payment(invoice_number:, amount:)
        Result.new(confirmed: true, transaction_reference: "STUB-#{invoice_number}")
      end
    end
  end
end
